xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/XWiki.java

```diff
@@ -2435,17 +2467,72 @@ public String getSkinFile(String filename, boolean forceSkinAction, XWikiContext
 
         // If all else fails, use the default base skin, even if the URLs could be invalid.
         URL url;
+
         if (forceSkinAction) {
             url = urlf.createSkinURL(filename, "skins", getDefaultBaseSkin(context), context);
         } else {
             url = urlf.createSkinURL(filename, getDefaultBaseSkin(context), context);
         }
+        queryParametersMap.putAll(getResourceURLCacheParameters(url));
+        url = urlf.addQueryParameters(url, queryParametersMap);
         return urlf.getURL(url, context);
     }
 
+    private String createURLForResource(Resource<?> resource, boolean forceSkinAction,
```

Comment:
As I keep saying, adding the parameters to the URL is Resource implementation's job because that's the only place where you really know what to add depending on the resource type. For example here you try to access the File as if all Resource were files on drive which is not true.

Reply:
I tried to answer to that comment by adding a query parameter when building the URL in EnvironmentSkinResource and by only returning the resource URL in getSkinFile when we already have a resource.
However when we only retrieve a path in getSkinFile I continue to try to retrieve the path to fill the query parameter, I did not change that.

I'm not sure it's what you had in mind but I've the feeling it would require far more refactoring to make it another way. I let you check.

Reply:
> I'm not sure it's what you had in mind

Looks good :)